ethash/ethash/ethash.cpp

```diff
@@ -283,24 +283,36 @@ bool verify_light(const hash256& header_hash, const hash256& mix_hash, uint64_t
     return is_less_or_equal(hash_final, boundary);
 }
 
-bool verify_full(const epoch_context& context, const hash256& header_hash, const hash256& mix_hash, uint64_t nonce,
-                 const hash256& boundary) noexcept {
-
+VerificationResult verify_full(const epoch_context& context, const hash256& header_hash, const hash256& mix_hash,
+                               uint64_t nonce, const hash256& boundary) noexcept {
     const hash512 hash_seed{detail::hash_seed(header_hash, nonce)};
     const hash256 hash_final{detail::hash_final(hash_seed, mix_hash)};
     if (!is_less_or_equal(hash_final, boundary)) {
-        return false;
+        return VerificationResult::kInvalidNonce;
     }
     const hash256 expected_mix_hash = detail::hash_mix(context, hash_seed);
-    return is_equal(mix_hash, expected_mix_hash);
+    if (!is_equal(mix_hash, expected_mix_hash)) {
+        return VerificationResult::kInvalidMixHash;
+    }
+    return VerificationResult::kOk;
+}
+
+VerificationResult verify_full(const uint64_t block_num, const hash256& header_hash, const hash256& mix_hash,
+                               uint64_t nonce, const hash256& boundary) noexcept {
+
+    static epoch_context_ptr epoch_context{nullptr, nullptr};
```

Comment:
This is not thread-safe.

Reply:
Yes but we actually don't need it to be.
All PoW verifications in Silkworm are single threaded